dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,411 @@
+    }
+
+    public KustoMemoryRecord(string key, string metadata, string? embedding, DateTime? timestamp = null)
+    {
+        this.Key = key;
+        this.Metadata = KustoSerializer.DeserializeMetadata(metadata);
+        this.Embedding = KustoSerializer.DeserializeEmbedding(embedding);
+        this.Timestamp = timestamp;
+    }
+
+    public MemoryRecord ToMemoryRecord()
+    {
+        return new MemoryRecord(this.Metadata, this.Embedding, this.Key, this.Timestamp);
+    }
+
+    public void WriteToCsvStream(CsvWriter streamWriter)
+    {
+        var jsonifiedMetadata = KustoSerializer.SerializeMetadata(this.Metadata);
+        var jsonifiedEmbedding = KustoSerializer.SerializeEmbedding(this.Embedding);
+        var isoFormattedDate = this.Timestamp?.ToString("o", CultureInfo.InvariantCulture) ?? string.Empty;
+
+        streamWriter.WriteField(this.Key);
+        streamWriter.WriteField(jsonifiedMetadata);
+        streamWriter.WriteField(jsonifiedEmbedding);
+        streamWriter.WriteField(isoFormattedDate);
+        streamWriter.CompleteRecord();
+    }
+}
+
+/// <summary>
+/// An implementation of <see cref="IMemoryStore"/> backed by a Kusto database.
+/// </summary>
+/// <remarks>The embedded data is saved to the Kusto database specified in the constructor.
+/// Similarity search capability is provided through a cosine similarity function (added on first collection creation). Use Kusto's "Table" to implement "Collection".
+/// </remarks>
+public class KustoMemoryStore : IMemoryStore
+{
+    private string _database;
+    private static ClientRequestProperties GetClientRequestProperties() => new()
+    {
+        Application = "SemanticKernel",
+    };
+
+    private readonly ICslQueryProvider _queryClient;
+    private readonly ICslAdminProvider _adminClient;
+
+    private static readonly ColumnSchema[] s_collectionColumns = new ColumnSchema[]
+    {
+        new ColumnSchema("Key", typeof(string).FullName),
+        new ColumnSchema("Metadata", typeof(object).FullName),
+        new ColumnSchema("Embedding", typeof(object).FullName),
+        new ColumnSchema("Timestamp", typeof(DateTime).FullName),
+    };
+
+    private const string c_collectionPrefix = "sk_memory_";
+    private static string GetTableName(string collectionName) => c_collectionPrefix + collectionName;
+    private static string GetCollectionName(string tableName) => tableName.Substring(c_collectionPrefix.Length);
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="KustoMemoryStore"/> class.
+    /// </summary>
+    /// <param name="kcsb">Kusto Connection String Builder.</param>
+    /// <param name="database">The database used for the tables.</param>
+    public KustoMemoryStore(KustoConnectionStringBuilder kcsb, string database)
+    {
+        this._database = database;
+        this._queryClient = KustoClientFactory.CreateCslQueryProvider(kcsb);
+        this._adminClient = KustoClientFactory.CreateCslAdminProvider(kcsb);
```

Comment:
I think we will need another constructor which will accept `ICslQueryProvider` and `ICslAdminProvider` as parameters. This will allow to add unit tests for this class.